### PR TITLE
Introduce an options reader for `package.json` files

### LINKED ISSUE #2112: typedoc doesn't recognize `package.json`, only `typedoc.json`
## Search terms

Running ``npm run docs`` for an npm script ``docs: typedoc src/index.ts`` does not read properties from ``package.json``:

```json
{
  // ...
  "typedocOptions": {
    "entryPoints": ["src/index.ts"],
    "out": "docs/typedocs",
    "includeVersion": true
  }
}
```

Only from ``typedoc.json``:

```json
{
   "entryPoints": ["src/index.ts"],
   "out": "docs/typedocs",
   "includeVersion": true
}
```

## Expected Behavior

read from ``package.json``

## Actual Behavior

reads only from ``typedoc.json``

## Steps to reproduce the bug

Run ``typedoc src/index.ts`` with the typedoc specs only in ``package.json``.

## Environment

-   Typedoc version: ``0.23.21``
-   TypeScript version: ``4.9.3``
-   Node.js version: ``v19.2.0``
-   OS: ``Ubuntu 20.04.5 LTS (Focal Fossa)``

### PR DESCRIPTION
**Pull request by jm4rtinez**
This PR introduces the capability to read TypeDoc options from `package.json` files. By default, the `PackageJsonReader` runs after the `TypeDocReader` but before the `TSConfigReader` so that users may specify a path to a `tsconfig.json` file.

Resolves #2112

### CODE REVIEW

File: src/lib/utils/options/readers/package-json.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        const absPath = resolve(cfgPath);
+
+        // Search for `package.json` first at the provided config path, or at
+        // `cfgPath/package.json` if the provided path is a directory.
+        const path = [absPath, join(absPath, "package.json")].find(
+            (p) => fs.existsSync(p) && fs.statSync(p).isFile()
+        );
+
+        if (!path) {
+            if (container.isSet("options")) {
```
Comment (Gerrit0) [inline src/lib/utils/options/readers/package-json.ts:28]: This shouldn't be an error - options specified the location of typedoc.json, for test projects I frequently have several typedoc.json files. I think we should just silently do nothing here.

File: src/lib/utils/options/readers/package-json.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                return;
+            }
+        } catch (err) {
+            ok(err instanceof Error);
+            this.logParsePkgFailure(logger, path);
+            return;
+        }
+
+        const opts: Record<string, any> | undefined = pkg["typedocOptions"];
+        if (opts === null || typeof opts !== "object") {
```
Comment (Gerrit0) [inline src/lib/utils/options/readers/package-json.ts:54]: It should not be an error if this key is not set, in that case, the user is likely configuring typedoc through another method.

**Review by Gerrit0 — CHANGES_REQUESTED** [review]
Thanks! Couple comments

File: src/lib/utils/options/readers/package-json.ts (hunk 1 above)
Comment (jm4rtinez) [reply src/lib/utils/options/readers/package-json.ts]: Gotcha, we'll carry on silently if a file path can't be resolved.

File: src/lib/utils/options/readers/package-json.ts (hunk 2 above)
Comment (jm4rtinez) [reply src/lib/utils/options/readers/package-json.ts:52]: Okay sounds good. We'll first check if the key exists or not, and only raise an error if the key is set but the data is invalid.

**Comment by jm4rtinez** [comment]
Thanks for the feedback! I've made some changes. Let me know if there's anything else.

**Review by Gerrit0 — APPROVED** [review]
I like it! This needs to wait for 0.24 to be merged, since it will otherwise conflict with the non-standard parsed by typedoc in packages mode from the package.json file, but it'll come in then!

**Pull request merged**